cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +237,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+func (s *PowerVSClusterScope) fetchResourceGroupID() (string, error) {
+	// TODO: Handle this case in wehbook
+	if s.ResourceGroup() == nil || s.ResourceGroup().Name == nil {
+		return "", fmt.Errorf("resource group name is not set")
+	}
+	rmv2, err := resourcemanagerv2.NewResourceManagerV2(&resourcemanagerv2.ResourceManagerV2Options{
+		Authenticator: s.session.Options.Authenticator,
+	})
+	if err != nil {
+		return "", err
+	}
+	if rmv2 == nil {
+		return "", fmt.Errorf("unable to get resource controller")
+	}
+	resourceGroup := s.ResourceGroup().Name
+	rmv2ListResourceGroupOpt := resourcemanagerv2.ListResourceGroupsOptions{Name: resourceGroup, AccountID: &s.session.Options.UserAccount}
+	resourceGroupListResult, _, err := rmv2.ListResourceGroups(&rmv2ListResourceGroupOpt)
+	if err != nil {
+		return "", err
+	}
+
+	if resourceGroupListResult != nil && len(resourceGroupListResult.Resources) > 0 {
+		rg := resourceGroupListResult.Resources[0]
+		resourceGroupID := *rg.ID
+		return resourceGroupID, nil
+	}
+
+	err = fmt.Errorf("could not retrieve resource group id for %s", *resourceGroup)
+	return "", err
+}
+
+// getVPCRegion returns region associated with VPC zone.
+func (s *PowerVSClusterScope) getVPCRegion() *string {
+	if s.IBMPowerVSCluster.Spec.VPC != nil {
+		return s.IBMPowerVSCluster.Spec.VPC.Region
+	}
+	// if vpc region is not set try to fetch corresponding region from power vs zone
+	if s.Zone() == nil {
+		return nil
+	}
+	region := endpoints.ConstructRegionFromZone(*s.Zone())
+	vpcRegion, err := genUtil.VPCRegionForPowerVSRegion(region)
+	if err != nil {
+		s.Error(err, fmt.Sprintf("failed to fetch vpc region associated with powervs region %s", region))
+		return nil
+	}
+	return &vpcRegion
+}
+
+// fetchVPCCRN returns VPC CRN.
+func (s *PowerVSClusterScope) fetchVPCCRN() (*string, error) {
+	vpcDetails, _, err := s.IBMVPCClient.GetVPC(&vpcv1.GetVPCOptions{
+		ID: s.GetVPCID(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return vpcDetails.CRN, nil
+}
+
+// fetchPowerVSServiceInstanceCRN returns Power VS service instance CRN.
+func (s *PowerVSClusterScope) fetchPowerVSServiceInstanceCRN() (*string, error) {
+	serviceInstanceID := s.GetServiceInstanceID()
+	pvsDetails, _, err := s.ResourceClient.GetResourceInstance(&resourcecontrollerv2.GetResourceInstanceOptions{
+		ID: &serviceInstanceID,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return pvsDetails.CRN, nil
+}
+
+// TODO(karthik-k-n): Decide on proper naming format for services.
+
+// GetServiceName returns name of given service type from spec or generate a name for it.
+func (s *PowerVSClusterScope) GetServiceName(resourceType infrav1beta2.ResourceType) *string {
+	switch resourceType {
+	case infrav1beta2.ResourceTypeServiceInstance:
+		if s.ServiceInstance() == nil || s.ServiceInstance().Name == nil {
+			return pointer.String(fmt.Sprintf("%s-serviceInstance", s.InfraCluster().GetName()))
+		}
+		return s.ServiceInstance().Name
+	case infrav1beta2.ResourceTypeNetwork:
```

Comment:
Fix this: Currently DHCP server creates network and its of some format as given here, If user provides name this wont work. I think dhcp server name cannot be set as we think. Explore!!

Reply:
Updated the logic to handle the scenarios when user provided DHCPserver name as well.